### DISCUSSION TITLE
MJPEG endpoint in /docs

### DISCUSSION THREAD

**Discussion by mariannasb** [fastapi/fastapi#6140] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def package_mjpeg(img_bytes: bytes, metadata: FrameMetadata = None) -> bytes:
        """Package jpeg bytes with MJPEG headers and frame separator

        Also append metadata if present
        """
        mjpeg = b"Content-Type: image/jpeg\r\n\r\n"
        mjpeg += img_bytes + b"\r\n"

        if metadata:
            mjpeg += b"Timestamp:" + bytes(str(metadata.time), "utf-8")
            mjpeg += b"\r\n"
            mjpeg += b"Sequence:" + bytes(str(metadata.sequence), "utf-8")
            mjpeg += b"\r\n"

        mjpeg += b"--frame\r\n"

        return mjpeg

async def async_gen(camera: BaseCamera) -> Generator[bytes, None, None]:
    """Video streaming async generator function."""
    yield b"--frame\r\n"
    try:
        while True:
            jpeg = camera.get_frame_jpeg()
            yield package_mjpeg(jpeg.data, jpeg.metadata)
            # need this sleep in order to be able to catch the disconnect
            await asyncio.sleep(0)
    except asyncio.CancelledError:
        app_logger.debug("Client disconnected")

@app.get("/mjpeg")
async def async_video_feed(dev: str = DEF_DEVICE) -> StreamingResponse:
    """Async video streaming route.

    Put this in the src attribute of an img tag.
    """
    try:
        cam = Camera(device=dev, stream=True)
    except V4L2Exception as e:
        raise APIV4L2Error(str(e)) from e
    except RuntimeError as e:
        raise APIInputError(str(e)) from e
    except Exception as e:
        raise APIGenError(
            f"{type(e).__name__} from {sys._getframe().f_code.co_name}(): {e}"
        ) from e

    return StreamingResponse(
        async_gen(cam),
        media_type="multipart/x-mixed-replace; boundary=frame",
    )
```

### Description

I have a MJPEG endpoint which works fine in the browser.

However it doesn't work (doesn't show the stream) inside the automatic docs endpoint (`<ip>:<port>/docs`).

If I look at the information it shows the response type as json so I suspect that is the problem.

I can change the response type shown by adding more information to the endpoint declaration like this (I have tried some different combinations in the content):
```
@app.get(
    "/mjpeg",
    responses={
        200: {
            "content": {
                "multipart/x-mixed-replace; boundary=frame": {},
                "image/jpeg": {},
            }
        },
    },
    response_class=StreamingResponse,
)
```
But nothing seems to make the stream show correctly in the docs.

Is that possible at all?
Or am I just missing some small detail?

### Operating System

Linux

### FastAPI Version

0.85.0

### Python Version

Python 3.10.5

**Comment by mbroton** [comment]
Hello!
I'm not sure about displaying a stream in the docs, but to declare a media type that you want, you could do this:
```py3
from fastapi import FastAPI
from fastapi.responses import StreamingResponse


app = FastAPI()


async def fake_video_streamer():
    for i in range(10):
        yield b"some fake video bytes"


class MyCustomResponse(StreamingResponse):
    media_type = "multipart/x-mixed-replace; boundary=frame"


@app.get("/mjpeg", response_class=MyCustomResponse)
async def async_video_feed():
    return MyCustomResponse(fake_video_streamer())
```
By doing this, the `openapi.json` will have correct information about the media type.
You can read more about this in FastAPI's documentation:
[StreamingResponse](https://fastapi.tiangolo.com/advanced/custom-response/#streamingresponse)
[Custom Response Class](https://fastapi.tiangolo.com/advanced/custom-response/#custom-response-class)

**Comment by mariannasb** [comment]
Thanks,

It doesn't really give me anything more than using `responses={...}`.
I get the media type information the same way and the stream still doesn't show in the docs page.

However it does look much nicer code-wise, so I'm definitely changing to use it instead of my solution.

I will try posting directly to swagger to see if it is something that is supported: https://github.com/swagger-api/swagger-ui/issues/8241

**Comment by tiangolo** [comment]
I would imagine it's not supported by Swagger UI but I'm not really sure. It would really belong in the Swagger repo.

Thanks for coming back to close this one here! ☕
